### PR TITLE
add clean-clone functionality under 'app' section

### PR DESCRIPTION
**Pull request by evanshortiss**
We need to be able to perform a clean clone using FHC to ensure out calls to stage/deploy are successfull from CI/CD. This should help us and others out.

### CODE REVIEW

**Comment by fh-automatron** [comment]
### Pull Request Quality Guidelines

As we continue to focus on quality, here are a few things to keep in mind:

* The code changes conform to the contributing & code style guides for this component (if they exist)
* A component owner should review your change (we've tagged these above)
* Automated testing has been added or updated for this change (e.g. UART, Unit/Acceptance tests)
* If updating server component this PR should be deployed to a grid somewhere (via `fhcap-cli`) - please provide a link to this
* Ensure JIRA is up to date before & after merge (e.g. deployment notes, tester notes, release notes, fix version)
* Any SonarQube issues should be addressed where applicable (use your own judgement)

Please ping a component owner (see `fh contrib`) when your builds are complete, changes are deployed and you're ready for a review.
Fhcap PR will be generated automatically.
To deploy your changes to a cluster place comment in this PR with pattern [x][y] where x=deploy and y=clustername you wish to deploy. 
You can also deploy your change manually using fhcap-cli

**Comment by evanshortiss** [comment]
@jasonmadigan perhaps you are the right person to ping (`fh contrib`) ?

Should I bump the version etc?

File: lib/cmd/fh3/app/clean-clone.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+var millicore = require("../../../utils/millicore.js");
+var fhreq = require("../../../utils/request");
+
+module.exports = cleanClone;
+
+cleanClone.usage = "fhc app clean-clone <appId>";
+cleanClone.desc = "Performs a clean clone of a git repo within the studio to resolve potential git issuse";
```
Comment (tagoh) [inline lib/cmd/fh3/app/clean-clone.js]: Please pass this string to i18n._() to enable localization. please see https://github.com/feedhenry/fh-fhc/commit/c5b56cc46392b199315b496fa7c5a02f43d90c99 for more details.

File: lib/cmd/fh3/app/clean-clone.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+function cleanClone (argv, cb) {
+  var appId = argv.appId;
+  var payload = {
+    clean: true
+  };
+
+  if (!appId) {
+    return cb(new Error('appId is required to perform a clean clone'));
```
Comment (tagoh) [inline lib/cmd/fh3/app/clean-clone.js]: ditto

File: lib/cmd/fh3/app/clean-clone.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  millicore.widgForAppId(appId, function (err, widgId) {
+    if (err) {
+      return cb(err);
+    }
+    //Sample URL: https://domain.feedhenry.com/box/api/connectors/{widget}/apps/{appId}/pull
+    fhreq.POST(fhreq.getFeedHenryUrl(), "/box/api/connectors/" + widgId + '/apps/' + appId + "/pull", payload, function (err, data) {
+      if (err) {
+        return cb(err);
+      } else if (!data.cacheKeys) {
+        return cb('Error in clean clone, unexpected response format');
```
Comment (tagoh) [inline lib/cmd/fh3/app/clean-clone.js]: ditto

File: lib/cmd/fh3/app/clean-clone.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      } else {
+        return cb(null, 'clean clone successful');
```
Comment (tagoh) [inline lib/cmd/fh3/app/clean-clone.js]: ditto

**Comment by evanshortiss** [comment]
@tagoh updated 👍

**Comment by evanshortiss** [comment]
@tagoh would it be possible to merge this and release an updated fh-fhc? This is blocking us on a deployment front.

File: lib/cmd/fh3/app/clean-clone.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      new Error(i18n._('appId is required to perform a clean clone'))
+    );
+  }
+
+  millicore.widgForAppId(appId, function (err, widgId) {
+    if (err) {
+      return cb(err);
+    }
+    //Sample URL: https://domain.feedhenry.com/box/api/connectors/{widget}/apps/{appId}/pull
+    fhreq.POST(fhreq.getFeedHenryUrl(), "/box/api/connectors/" + widgId + '/apps/' + appId + "/pull", payload, function (err, data) {
```
Comment (jasonmadigan) [inline lib/cmd/fh3/app/clean-clone.js]: Switch to use `/box/api/projects/` instead, `/box/api/connectors/` is for services, although I believe both will work.

**Comment by jasonmadigan** [comment]
@evanshortiss Some `fhc` updates are pending the release of `3.12`, which isn't due for release until the end of the month. To unblock in the meantime, I'd point `package.json`'s at a branch.

**Comment by shannonmpoole** [comment]
@evanshortiss @jasonmadigan Wouldn't it make more sense to add a `--clean` flag to the existing `fhc git pull` command? Versus creating a whole new command.

**Comment by evanshortiss** [comment]
@shannonmpoole @jasonmadigan I'm open to suggestion. I placed it in a new command since technically it's not a "clean" pull, but a complete clone, and separate studio option.

@jasonmadigan will update ASAP with new URL. Our plan in the meantime is what you've suggested, thanks 👍

**Comment by jasonmadigan** [comment]
@wei-lee / @maleck13 - any thoughts on adding this to the existing `fhc git pull` vs a new command as @evanshortiss has added?

**Comment by wei-lee** [comment]
I think it makes more sense to add a new flag to the existing command. Just so that users won't be confused about when to use which command.

File: lib/cmd/common/git.js [hunk 6]
```diff
@@ -26,9 +26,12 @@ function git(argv, cb) {
   if (action === 'props') {
     appId = fhc.appId(args[1]);
     return props(appId, cb);
-  } else if (action === 'pull') {
+  } else if (action === 'pull' && !argv.clean) {
     appId = fhc.appId(args[1]);
-    return pull(appId, cb);
+    return pull(appId, argv.clean, cb);
```
Comment (fh-qube) [inline lib/cmd/common/git.js]: BLOCKER "pull" expects "2" arguments, but "3" were provided. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS930)

**Comment by fh-qube** [comment]
SonarQube analysis reported 1 issue:
* BLOCKER 1 blocker

Watch the comments in this conversation to review them.

**Comment by evanshortiss** [comment]
@wei-lee @jasonmadigan @shannon updated to add "--clean" flag for git pull.

File: package.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "fh-fhc",
   "description": "A Command Line Interface for FeedHenry",
-  "version": "2.11.1-BUILD-NUMBER",
+  "version": "2.12.0-BUILD-NUMBER",
```
Comment (evanshortiss) [inline package.json]: Bumped to 2.12 since I had to merge from master which  was 2.11.1

**Comment by evanshortiss** [comment]
After using this, I'm thinking --clean will need to do two things. First call the new _cleanClone_ function, then call the existing _pull_ function.

**Comment by evanshortiss** [comment]
Have confirmed previous comment, and implemented it.

**Comment by wei-lee** [comment]
@evanshortiss sorry about the delay, i am going to merge this back to master.  Can you please rebase and squash the commits? I will then verify and merge the PR.

Thanks.

**Comment by evanshortiss** [comment]
@wei-lee happy to do the rebase. It's worth noting that GitHub has the awesome new "Squash and Merge" feature, but will squash manually 👍

**Comment by wei-lee** [comment]
@evanshortiss I have verified the change and it works fine. However, not sure if you have done rebase already? For some reason, there are no squash commits option on this page. If could be because you merged master branch into the feature branch, but not rebased?

**Pull request merged**